pepper-apis/src/main/java/org/broadinstitute/ddp/content/I18nContentRenderer.java

```diff
@@ -40,6 +42,32 @@ private static Long getDefaultLanguageId(Handle handle) {
         }
     }
 
+    public static RenderValueProvider newValueProvider(Handle handle, long participantUserId) {
+        return newValueProvider(handle, participantUserId, new HashMap<>());
+    }
+
+    public static RenderValueProvider newValueProvider(Handle handle, long participantUserId, Map<String, String> snapshot) {
+        var builder = new RenderValueProvider.Builder();
+        builder.setDate(LocalDate.now());
+
+        UserProfile profile = handle.attach(UserProfileDao.class)
+                .findProfileByUserId(participantUserId)
+                .orElse(null);
+        if (profile != null) {
+            if (profile.getFirstName() != null) {
+                builder.setParticipantFirstName(profile.getFirstName());
+            }
+            if (profile.getLastName() != null) {
+                builder.setParticipantLastName(profile.getLastName());
+            }
+        }
+
+        // Override with saved snapshot substitutions.
```

Comment:
You could probably make this comment more helpful by adding that the saved snapshot substitutions are for maintaining consistency for the user by keeping track of the value that was substituted when the user filled out the survey to show when they view the survey after they submit it when the study is configured to do so.  It's possible to piece that information together, but knowing that makes it easier to understand the code.

Reply:
done